### ISSUE TITLE
TypeError: google.visualization.DataTable is not a constructor 

### ISSUE THREAD

**Issue by TheTechnoMan** [kevinkhill/lavacharts#81]
Something seems to be very weird since a couple of days. I randomly get:

TypeError: google.visualization.DataTable is not a constructor
 graphs:871:26

And therefore the Charts won't load. It is totally random, sometimes it works fine, simetimes the error occurres.

Warm regards!

**Comment by kevinkhill** [comment]
I think you, and others have found a race condition. If you look at #75 we had quite a discussion about it, and a fix in the works.

**Comment by TheTechnoMan** [comment]
Oh okay, thanks :)

**Comment by TheTechnoMan** [comment]
Wait, I am a bit confused. THe issue is closed, but has it been fixed?

**Comment by kevinkhill** [comment]
Well, I believe I fixed the race condition problem. I'm away from home right now and don't remember if I merged the fix yet.

**Comment by TheTechnoMan** [comment]
Hmm, I ran a composer update but everything is up to date, so I guess you did not merge it yet.

**Comment by kevinkhill** [comment]
OK, so the fix is merged into the 3.0-unstable branch (were you on 2.5?)

I am working on the docs for the release of 3.0 (hopefully soon, I've been dragging my feet) but I have pretty much locked the API of the 3.0 branch so depending on unstable will eventually lead to the actual 3.0 release

**Comment by TheTechnoMan** [comment]
I'm on "khill/lavacharts": "3.0.x-dev",

**Comment by kevinkhill** [comment]
Ahhh, yes, I didn't merge the fix into the stable branch because I didn't get to test it as much as I wanted. We just confirmed it worked for @SendDerek.

**Comment by kevinkhill** [comment]
If you want to help me out and test against the unstable branch, that would help me be more comfortable with merging my changes into the stable branch.

**Comment by TheTechnoMan** [comment]
Sorry if that sounds dumb, but doesn't dev mean that I am on the unstable one?

**Comment by TheTechnoMan** [comment]
nvm. I get it. Okay I will test it. :)

**Comment by kevinkhill** [comment]
It's not dumb, I thought the same thing until I learned that: composer labels *branches* as branch-dev but just numbers refer to tags in branches. This is why I have the 3.0 branch and the 3.0-unstable branch; one to work in and one with stable code.

**Comment by TheTechnoMan** [comment]
How would I require it? "khill/lavacharts": "3.0.x-unstable", does not work.

Thanks

**Comment by kevinkhill** [comment]
I believe it would be "3.0-unstable-dev"

**Comment by TheTechnoMan** [comment]
Okay. Sadly it is not fixed. After loading it 3 times the error occurred again :(

**Comment by kevinkhill** [comment]
Damn! I thought for sure I covered my bases to prevent this. Can you take a screen shot or copy from the dev console what Lavacharts is outputting into the page? The script tags specifically.

**Comment by TheTechnoMan** [comment]
Sure.

```
                      <script type="text/javascript">/**
 * jvent.min.js
 *
 * Author:   Guille Paz
 * Email:    guille87paz@gmail.com
 * Github:   https://github.com/pazguille/jvent
 * Homepage: http://pazguille.me
 * License:  MIT
 */
!function(t){"use strict";function e(){}e.prototype.on=function(t,e){return this._collection=this._collection||{},this._collection[t]=this._collection[t]||[],this._collection[t].push(e),this},e.prototype.once=function(t,e){function i(){o.off(t,i),e.apply(this,arguments)}var o=this;return i.listener=e,this.on(t,i),this},e.prototype.off=function(t,e){var i=this._collection[t],o=0;if(void 0!==i)for(o;o<i.length;o+=1)if(i[o]===e||i[o].listener===e){i.splice(o,1);break}return 0===i.length&&this.removeAllListeners(t),this},e.prototype.removeAllListeners=function(t){return this._collection=this._collection||{},delete this._collection[t],this},e.prototype.listeners=function(t){return this._collection=this._collection||{},this._collection[t]},e.prototype.emit=function(){if(void 0===this._collection)return this;var t,e=[].slice.call(arguments,0),i=e.shift(),o=this._collection[i],n=0;if(o)for(o=o.slice(0),t=o.length,n;t>n;n+=1)o[n].apply(this,e);return this},"function"==typeof t.define&&void 0!==t.define.amd?t.define("Jvent",[],function(){return e}):"undefined"!=typeof module&&void 0!==module.exports?module.exports=e:t.Jvent=e}(this);

/**
 * lava.js
 *
 * Author:  Kevin Hill
 * Email:   kevinkhill@gmail.com
 * Github:  https://github.com/kevinkhill/lavacharts
 * License: MIT
 */
var lava = lava || {};

(function() {
  "use strict";

  this.charts            = [];
  this.dashboards        = [];
  this.registeredCharts  = [];
  this.readyCallback     = null;
  this.renderDashboards  = null;

  this.events = new Jvent();

  //var registeredActions = [];

  /**
   * LavaChart object.
   *
   * @constructor
   */
  this.Chart = function() {
    var self = this;
    this.data    = null;
    this.chart   = null;
    this.options = null;
    this.formats = [];
    this.render  = function(){};
    this.setData = function(){};
    this.redraw  = function(){};
    this.applyFormats = function (formatArr) {
      for(var a=0; a < formatArr.length; a++) {
        var formatJson = formatArr[a];
        var formatter = new google.visualization[formatJson.type](formatJson.config);
            formatter.format(self.data, formatJson.index);
      }
    };
  };

  /**
   * Dashboard object.
   *
   * @constructor
   */
  this.Dashboard = function() {
    this.render    = null;
    this.data      = null;
    this.bindings  = [];
    this.dashboard = null;
    this.callbacks = [];
  };

  this.Callback = function (label, func) {
    this.label = label;
    this.func  = func;
  };

  this.ready = function (callback) {
    if (typeof callback !== 'function') {
      throw new Error('[Lavacharts] ' + typeof callback + ' is not a valid callback.');
    }

    lava.readyCallback = callback;
  };

  /**
   * Event wrapper for chart events.
   *
   *
   * Used internally when events are applied so the user event function has
   * access to the chart within the event callback.
   *
   * @param {object} event
   * @param {object} chart
   * @param {function} callback
   */
  this.event = function (event, chart, callback) {
    return callback(event, chart);
  };

  /**
   * Registers a chart as being on screen, accessible to redraws.
   */
  this.registerChart = function(type, label) {
    this.registeredCharts.push(type + ':' + label);
  };

  /**
   * Loads a new DataTable into the chart and redraws.
   *
   *
   * Used with an AJAX call to a PHP method returning DataTable->toJson(),
   * a chart can be dynamically update in page, without reloads.
   *
   * @param {string} chartLabel
   * @param {string} json
   * @param {function} callback
   */
  this.loadData = function (chartLabel, json, callback) {
    lava.getChart(chartLabel, function (chart, LavaChart) {
      if (typeof json.data != 'undefined') {
        LavaChart.setData(json.data);
      } else {
        LavaChart.setData(json);
      }

      if (typeof json.formats != 'undefined') {
        LavaChart.applyFormats(json.formats);
      }

      LavaChart.redraw();

      if (typeof callback == 'function') {
        callback(LavaChart.chart);
      }
    });
  };

  this.getDashboard = function (label, callback) {
    if (typeof lava.dashboards[label] === 'undefined') {
      throw new Error('[Lavacharts] Dashboard "' + label + '" was not found.');
    }

    var lavaDashboard = lava.dashboards[label];

    if (typeof callback !== 'function') {
      throw new Error('[Lavacharts] ' + typeof callback + ' is not a valid callback.');
    }

    callback(lavaDashboard.dashboard, lavaDashboard);
  };

  /**
   * Returns the GoogleChart and the LavaChart objects
   *
   *
   * The GoogleChart object can be used to access any of the available methods such as
   * getImageURI() or getChartLayoutInterface().
   * See https://google-developers.appspot.com/chart/interactive/docs/gallery/linechart#methods
   * for some examples relative to LineCharts.
   *
   * The LavaChart object holds all the user defined properties such as data, options, formats,
   * the google chart object, and relative methods for internal use.
   *
   * Just to clarify:
   *  - The first returned callback value is a property of the LavaChart.
   *    It was add as a shortcut to avoid chart.chart to access google's methods of the chart.
   *
   *  - The second returned callback value is the LavaChart, which holds the GoogleChart and other
   *    important information. It was added to not restrict the user to only getting the GoogleChart
   *    returned, and as the second value because it is less useful / rarely accessed.
   *
   * @param  {string}   chartLabel
   * @param  {function} callback
   */
  this.getChart = function (chartLabel, callback) {
    if (typeof chartLabel != 'string') {
      throw new Error('[Lavacharts] ' + typeof chartLabel + ' is not a valid chart label.');
    }

    if (typeof callback != 'function') {
      throw new Error('[Lavacharts] ' + typeof callback + ' is not a valid callback.');
    }

    var chartTypes = Object.keys(lava.charts);
    var LavaChart;

    var search = chartTypes.some(function (type) {
      if (typeof lava.charts[type][chartLabel] !== 'undefined') {
        LavaChart = lava.charts[type][chartLabel];

        return true;
      } else {
        return false;
      }
    });

    if (search === false) {
      throw new Error('[Lavacharts] Chart "' + chartLabel + '" was not found.');
    }

    callback(LavaChart.chart, LavaChart);
  };

  /**
   * Redraws all of the registed charts on screen.
   *
   *
   * This method is attached to the window resize event with a 300ms debounce
   * to make the charts responsive to the browser resizing.
   */
  this.redrawCharts = function() {
    var timer, delay = 300;

    clearTimeout(timer);

    timer = setTimeout(function() {
      for(var c = 0; c < lava.registeredCharts.length; c++) {
        var parts = lava.registeredCharts[c].split(':');

        lava.charts[parts[0]][parts[1]].redraw();
      }
    }, delay);
  };

  this.run = function () {
    var s = document.createElement('script');
        s.type = 'text/javascript';
        s.src = '//www.google.com/jsapi';
        s.onload = s.onreadystatechange = function (event) {
          event = event || window.event;

          if (event.type === "load" || (/loaded|complete/.test(this.readyState))) {
            this.onload = this.onreadystatechange = null;
            
            lava.events.emit('jsapi:ready');
          }
        };

    document.body.appendChild(s);
  };

}).apply(lava);

/**
 * Adding the resize event listener for redrawing charts.
 */
window.addEventListener("resize", window.lava.redrawCharts);

lava.run();
</script>
<script type="text/javascript">
        lava.events.on('jsapi:ready', function() {
            /**
             * If the object does not exist for a given chart type, initialize it.
             * This will prevent overriding keys when multiple charts of the same
             * type are being rendered on the same page.
             */
            if ( typeof lava.charts.AreaChart == "undefined" ) {
                lava.charts.AreaChart = {};
            }

            //Creating a new lavachart object
            lava.charts.AreaChart["registrations"] = new lava.Chart();

            //Checking if output div exists
            if (! document.getElementById("registrations_div")) {
                throw new Error('[Lavacharts] No matching element was found with ID "registrations_div"');
            }

            lava.charts.AreaChart["registrations"].render = function (data) {
                var $this = lava.charts.AreaChart["registrations"];

                $this.data = new google.visualization.DataTable({"cols":[{"type":"string","label":"Date"},{"type":"number","label":"Activated"},{"type":"number","label":"Not activated"}],"rows":[]}, 0.6);

                $this.options = {"backgroundColor":{"fill":"#333333","stroke":"#868686","strokeWidth":3},"hAxis":{"textStyle":{"color":"#bbb"},"gridlines":{"color":"#bbb","count":7}},"vAxis":{"textStyle":{"color":"#bbb"},"gridlines":{"color":"#646464","count":5}},"lineWidth":4,"legend":{"position":"top","alignment":"center","textStyle":{"color":"#bbb"}},"chartArea":{"width":"90%"},"titleTextStyle":{"color":"#bbb"},"focusTarget":"category"};

                $this.chart = new google.visualization.AreaChart(document.getElementById("registrations_div"));

                

                

                $this.chart.draw($this.data, $this.options);
            };

            lava.charts.AreaChart["registrations"].setData = function (data) {
                var $this = lava.charts.AreaChart["registrations"];

                $this.data = new google.visualization.DataTable(data, 0.6);
            };

            lava.charts.AreaChart["registrations"].redraw = function () {
                var $this = lava.charts.AreaChart["registrations"];

                $this.chart.draw($this.data, $this.options);
            };

            lava.registerChart("AreaChart", "registrations");

            google.load('visualization', '1', {
                packages: ['corechart'],
                callback: function() {
                    lava.charts.AreaChart["registrations"].render();
                    //lava.readyCallback();
                }
            });
        });
</script>            
                        
<script type="text/javascript">
        lava.events.on('jsapi:ready', function() {
            /**
             * If the object does not exist for a given chart type, initialize it.
             * This will prevent overriding keys when multiple charts of the same
             * type are being rendered on the same page.
             */
            if ( typeof lava.charts.AreaChart == "undefined" ) {
                lava.charts.AreaChart = {};
            }

            //Creating a new lavachart object
            lava.charts.AreaChart["loginAttempts"] = new lava.Chart();

            //Checking if output div exists
            if (! document.getElementById("loginAttempts_div")) {
                throw new Error('[Lavacharts] No matching element was found with ID "loginAttempts_div"');
            }

            lava.charts.AreaChart["loginAttempts"].render = function (data) {
                var $this = lava.charts.AreaChart["loginAttempts"];

                $this.data = new google.visualization.DataTable({"cols":[{"type":"string","label":"Date"},{"type":"number","label":"Successful Attempts"},{"type":"number","label":"Failed Attempts"},{"type":"number","label":"Failed because not activated"}],"rows":[{"c":[{"v":"November 21"},{"v":0},{"v":0},{"v":0}]},{"c":[{"v":"November 22"},{"v":1},{"v":0},{"v":0}]},{"c":[{"v":"November 23"},{"v":1},{"v":0},{"v":0}]},{"c":[{"v":"November 24"},{"v":3},{"v":0},{"v":0}]},{"c":[{"v":"November 25"},{"v":1},{"v":0},{"v":0}]},{"c":[{"v":"November 26"},{"v":1},{"v":0},{"v":0}]},{"c":[{"v":"November 27"},{"v":2},{"v":0},{"v":0}]}]}, 0.6);

                $this.options = {"backgroundColor":{"fill":"#333333","stroke":"#868686","strokeWidth":3},"hAxis":{"textStyle":{"color":"#bbb"},"gridlines":{"color":"#bbb","count":7}},"vAxis":{"textStyle":{"color":"#bbb"},"gridlines":{"color":"#646464","count":5}},"lineWidth":4,"legend":{"position":"top","alignment":"center","textStyle":{"color":"#bbb"}},"chartArea":{"width":"90%"},"titleTextStyle":{"color":"#bbb"},"focusTarget":"category"};

                $this.chart = new google.visualization.AreaChart(document.getElementById("loginAttempts_div"));

                

                

                $this.chart.draw($this.data, $this.options);
            };

            lava.charts.AreaChart["loginAttempts"].setData = function (data) {
                var $this = lava.charts.AreaChart["loginAttempts"];

                $this.data = new google.visualization.DataTable(data, 0.6);
            };

            lava.charts.AreaChart["loginAttempts"].redraw = function () {
                var $this = lava.charts.AreaChart["loginAttempts"];

                $this.chart.draw($this.data, $this.options);
            };

            lava.registerChart("AreaChart", "loginAttempts");

            google.load('visualization', '1', {
                packages: ['corechart'],
                callback: function() {
                    lava.charts.AreaChart["loginAttempts"].render();
                    //lava.readyCallback();
                }
            });
        });
</script>            
```

**Comment by kevinkhill** [comment]
We believed that the issue was my lava.js module running before the Google jsapi loaded. I wrapped up the module to run after getting an event that the jsapi loaded.

**Comment by TheTechnoMan** [comment]
Hmm, seems like something is still wrong, even though that your approach seems to be understandable.

**Comment by kevinkhill** [comment]
Me neither, I just kind hack away with basic knowledge until it works. I'm sure there is a more elegant way to do the lava.js module.

I had the idea once before about downloading the jsapi module myself and including it in the lavacharts output. The downside being, not getting any updates and having to manually update the jsapi myself. The upside being that I wouldn't have to track or worry about the jsapi module loading and being available before running my lava.js

**Comment by TheTechnoMan** [comment]
To be honest, I personally would be happy about that. For example, if you work on a host or internal network without internet access, you won't be able to access the charts. Would that work if you include it locally? I think you once told me that it still loads other frmeote files, right?

**Comment by kevinkhill** [comment]
Oh yeah, I forgot about that... 
The jsapi initially loads, then dynamically loads whichever charts are defined. 

Which you just led me to think, maybe me catching the jsapi loading isn't enough to stop the race condition of the charts loading??

Idk, what's difficult is that I personally haven't ran into this issue before, so it's hard to debug.  How many charts are you trying to load on one page?

**Comment by TheTechnoMan** [comment]
Hmm, might be. Currently 2.

**Comment by SendDerek** [comment]
Throwing my experience into the ring...  I'm rendering four different charts using ajax.  I'm also using memcached to help speed things up.  It's when things are cached is when I get one or both of the error messages as mentioned in https://github.com/kevinkhill/lavacharts/issues/75#issuecomment-158795135.

To test the issue @TheTechnoMan is reporting, I updated `composer.json` with `"khill/lavacharts": "3.0-unstable-dev"` and ran `composer update khill/lavacharts`.

    Loading composer repositories with package information
    Updating dependencies (including require-dev)
      - Removing khill/lavacharts (3.0.x-dev a65bcda)
      - Installing khill/lavacharts (dev-3.0-unstable 820d7a7)
        Downloading: 100%         

Sweet.  Now, it gets strange for me too.  I can confirm that in my development environment, the race condition appears to be fixed.  I do not get any error messages and I don't need to introduce a `sleep(1)` call anywhere.  However, in production, the race condition is still a problem as I see the following error for all four charts I'm attempting to render:

    Uncaught TypeError: google.visualization.DataTable is not a function

Notice it's not exactly the same error message as @TheTechnoMan is getting.

Adding the `sleep(1)` call makes things work again without any errors.

I'd love to help provide complete source code and my environment where the problem persists, but we'd need to deal with NDAs and what-not.  I'm up for going through a debug session over Google Hangouts or something, but we'd still need to do an NDA for that.  My best advice for how to reproduce is below, but I can't say whether or not it's minimum code to reproduce:

1.) Setup charts to be populated with ajax calls using `lava.loadData()` for ajax in your view:

    <body>
        {!! \Lava::render('ComboChart', 'sessions', 'sessions') !!}
    </body>
    
    <script>
    function getDataAndLoadLava(url, id1) {
        $.getJSON(url, function(data) {
            lava.loadData(id1, data);
        });
    }

    getDataAndLoadLava('http://example.com/data/sessions', 'sessions');
    </script>

2.) Put the JSON responses that `lava.loadData()` fetches into memcached.  In this example, json data is coming from url *http://example.com/data/sessions*:

    /**
     * Responds to requests to GET /data/sessions
     */
    public function getSessions()
    {
        return Cache::remember("sessions", 720, function()  {
            return $this->getSessions()->toJson();
        });
    }

3.) In the index or page controller, load the table with an empty table.

        $sessionsComboChart = Lava::ComboChart('Sessions', Lava::DataTable(), [
            'title' => "Sessions",
        ]);

        return view('index', compact('sessionComboChart'));

When the table is first loaded, before the `lava.loadData()` line is called in the javascript, you may see a message like this:
![image](https://cloud.githubusercontent.com/assets/469263/11574387/a1b6612a-99c8-11e5-92f9-7fb738bee7a4.png)

4.) Get to a point where you see the chart being drawn on the webpage.  You may need to introduce a `sleep(2)` call or whatever to emulate going out to some service and fetching data.  For this first request, the 'sessions' key was not yet stored in memcached, so we simply refresh the page to make use of the JSON in memcached.  At this point, you may notice the race condition error described above (DataTable is not a function).

Let me know if you're interested in any other debugging information and I'll do my best to provide it to you.  I hope this helps.  Sounds like a tricky issue.

There's both good and bad to including the jsapi source into your own project.  One of my first thoughts was that I would rather see you spending time on implementing new charts or features than maintaining jsapi package every time it's updated.  I can also see the argument for having it included just in case there's an offline application.

**Comment by kevinkhill** [comment]
That's a lot of info, so, here's my thoughts... 
#3, the error is from loading the empty data table which has no columns defined. (You probably figured that out) while testing, I would   create the DataTable structure to give the chart, to remove that error, and later add the rows with the ajax. So maybe define and store the DataTable structures in a central place, so can throw it in the chart, and then populate with your DB.

As for memcache, I've never even used it, so I don't really know what effect that has on lavacharts. I should probably read up and start debugging with it.

**Comment by kevinkhill** [comment]
I just gave myself an idea for a DataTable method. One to clone and strip off the rows, or hide them, for the purpose of ajax renders. So instead of passing the ```Lava::DataTable()``` you could maybe use a method like ```bare()``` or ```hideRows()``` to avoid the No Columns error

**Comment by SendDerek** [comment]
Yeah, the No Columns error is expected.  I forgot to mention that.  It'd be cool to have a bare DataTable to avoid that message.

Memcached is very easy to get running in Laravel.  Take a look at the documentation (http://laravel.com/docs/5.1/cache).

Suggest a time to do a Google Hangout if you do find the time.  We'll try to work something out. :-)

**Comment by kevinkhill** [comment]
Version 3.0 is here! Hopefully all these lingering issues are resolved. Fingers crossed. As always, feel free to open new issues if they arise :+1:

**Comment by SendDerek** [comment]
That's excellent news!  Congrats. :+1:

**Comment by kevinkhill** [comment]
I realized I was dragging my feet and it really was ready to go. Just that whole releasing-my-creation-into-the-wild anxiety, haha. So, bite the bullet, release, and solve issues as they arise.

**Comment by TheTechnoMan** [comment]
Sadly still not fixed:

TypeError: google.visualization.DataTable is not a constructor
graphs (Line 690, Column 30)

:(

**Comment by kevinkhill** [comment]
Can you show me more? Like where in the javascript output this is happening? I haven't been getting this error in my tests, so it's hard to debug

**Comment by TheTechnoMan** [comment]
Hey,

ths is the JS part:

```
                      <script type="text/javascript">/**
 * jvent.min.js
 *
 * Author:   Guille Paz
 * Email:    guille87paz@gmail.com
 * Github:   https://github.com/pazguille/jvent
 * Homepage: http://pazguille.me
 * License:  MIT
 */
!function(t){"use strict";function e(){}e.prototype.on=function(t,e){return this._collection=this._collection||{},this._collection[t]=this._collection[t]||[],this._collection[t].push(e),this},e.prototype.once=function(t,e){function i(){o.off(t,i),e.apply(this,arguments)}var o=this;return i.listener=e,this.on(t,i),this},e.prototype.off=function(t,e){var i=this._collection[t],o=0;if(void 0!==i)for(o;o<i.length;o+=1)if(i[o]===e||i[o].listener===e){i.splice(o,1);break}return 0===i.length&&this.removeAllListeners(t),this},e.prototype.removeAllListeners=function(t){return this._collection=this._collection||{},delete this._collection[t],this},e.prototype.listeners=function(t){return this._collection=this._collection||{},this._collection[t]},e.prototype.emit=function(){if(void 0===this._collection)return this;var t,e=[].slice.call(arguments,0),i=e.shift(),o=this._collection[i],n=0;if(o)for(o=o.slice(0),t=o.length,n;t>n;n+=1)o[n].apply(this,e);return this},"function"==typeof t.define&&void 0!==t.define.amd?t.define("Jvent",[],function(){return e}):"undefined"!=typeof module&&void 0!==module.exports?module.exports=e:t.Jvent=e}(this);

/**
 * lava.js
 *
 * Author:  Kevin Hill
 * Email:   kevinkhill@gmail.com
 * Github:  https://github.com/kevinkhill/lavacharts
 * License: MIT
 */
var lava = lava || {};

(function() {
  "use strict";

  this.charts            = [];
  this.dashboards        = [];
  this.registeredCharts  = [];
  this.readyCallback     = null;
  this.renderDashboards  = null;

  this.events = new Jvent();

  //var registeredActions = [];

  /**
   * LavaChart object.
   *
   * @constructor
   */
  this.Chart = function() {
    var self = this;
    this.data    = null;
    this.chart   = null;
    this.options = null;
    this.formats = [];
    this.render  = function(){};
    this.setData = function(){};
    this.redraw  = function(){};
    this.applyFormats = function (formatArr) {
      for(var a=0; a < formatArr.length; a++) {
        var formatJson = formatArr[a];
        var formatter = new google.visualization[formatJson.type](formatJson.config);
        formatter.format(self.data, formatJson.index);
      }
    };
  };

  /**
   * Dashboard object.
   *
   * @constructor
   */
  this.Dashboard = function() {
    this.render    = null;
    this.data      = null;
    this.bindings  = [];
    this.dashboard = null;
    this.callbacks = [];
  };

  this.Callback = function (label, func) {
    this.label = label;
    this.func  = func;
  };

  this.ready = function (callback) {
    if (typeof callback !== 'function') {
      throw new Error('[Lavacharts] ' + typeof callback + ' is not a valid callback.');
    } else {
      var callback = function(){};
    }

    lava.readyCallback = callback;
  };

  /**
   * Event wrapper for chart events.
   *
   *
   * Used internally when events are applied so the user event function has
   * access to the chart within the event callback.
   *
   * @param {object} event
   * @param {object} chart
   * @param {function} callback
   */
  this.event = function (event, chart, callback) {
    return callback(event, chart);
  };

  /**
   * Registers a chart as being on screen, accessible to redraws.
   */
  this.registerChart = function(type, label) {
    this.registeredCharts.push(type + ':' + label);
  };

  /**
   * Loads a new DataTable into the chart and redraws.
   *
   *
   * Used with an AJAX call to a PHP method returning DataTable->toJson(),
   * a chart can be dynamically update in page, without reloads.
   *
   * @param {string} chartLabel
   * @param {string} json
   * @param {function} callback
   */
  this.loadData = function (chartLabel, json, callback) {
    lava.getChart(chartLabel, function (chart, LavaChart) {
      if (typeof json.data != 'undefined') {
        LavaChart.setData(json.data);
      } else {
        LavaChart.setData(json);
      }

      if (typeof json.formats != 'undefined') {
        LavaChart.applyFormats(json.formats);
      }

      LavaChart.redraw();

      if (typeof callback == 'function') {
        callback(LavaChart.chart);
      }
    });
  };

  this.getDashboard = function (label, callback) {
    if (typeof lava.dashboards[label] === 'undefined') {
      throw new Error('[Lavacharts] Dashboard "' + label + '" was not found.');
    }

    var lavaDashboard = lava.dashboards[label];

    if (typeof callback !== 'function') {
      throw new Error('[Lavacharts] ' + typeof callback + ' is not a valid callback.');
    }

    callback(lavaDashboard.dashboard, lavaDashboard);
  };

  /**
   * Returns the GoogleChart and the LavaChart objects
   *
   *
   * The GoogleChart object can be used to access any of the available methods such as
   * getImageURI() or getChartLayoutInterface().
   * See https://google-developers.appspot.com/chart/interactive/docs/gallery/linechart#methods
   * for some examples relative to LineCharts.
   *
   * The LavaChart object holds all the user defined properties such as data, options, formats,
   * the google chart object, and relative methods for internal use.
   *
   * Just to clarify:
   *  - The first returned callback value is a property of the LavaChart.
   *    It was add as a shortcut to avoid chart.chart to access google's methods of the chart.
   *
   *  - The second returned callback value is the LavaChart, which holds the GoogleChart and other
   *    important information. It was added to not restrict the user to only getting the GoogleChart
   *    returned, and as the second value because it is less useful / rarely accessed.
   *
   * @param  {string}   chartLabel
   * @param  {function} callback
   */
  this.getChart = function (chartLabel, callback) {
    if (typeof chartLabel != 'string') {
      throw new Error('[Lavacharts] ' + typeof chartLabel + ' is not a valid chart label.');
    }

    if (typeof callback != 'function') {
      throw new Error('[Lavacharts] ' + typeof callback + ' is not a valid callback.');
    }

    var chartTypes = Object.keys(lava.charts);
    var LavaChart;

    var search = chartTypes.some(function (type) {
      if (typeof lava.charts[type][chartLabel] !== 'undefined') {
        LavaChart = lava.charts[type][chartLabel];

        return true;
      } else {
        return false;
      }
    });

    if (search === false) {
      throw new Error('[Lavacharts] Chart "' + chartLabel + '" was not found.');
    }

    callback(LavaChart.chart, LavaChart);
  };

  /**
   * Redraws all of the registered charts on screen.
   *
   *
   * This method is attached to the window resize event with a 300ms debounce
   * to make the charts responsive to the browser resizing.
   */
  this.redrawCharts = function() {
    var timer, delay = 300;

    clearTimeout(timer);

    timer = setTimeout(function() {
      for(var c = 0; c < lava.registeredCharts.length; c++) {
        var parts = lava.registeredCharts[c].split(':');

        lava.charts[parts[0]][parts[1]].redraw();
      }
    }, delay);
  };

  this.run = function () {
    var s = document.createElement('script');
    s.type = 'text/javascript';
    s.src = '//www.google.com/jsapi';
    s.onload = s.onreadystatechange = function (event) {
      event = event || window.event;

      if (event.type === "load" || (/loaded|complete/.test(this.readyState))) {
        this.onload = this.onreadystatechange = null;

        lava.events.emit('jsapi:ready', window.google);
      }
    };

    document.head.appendChild(s);
  };

}).apply(lava);

/**
 * Adding the resize event listener for redrawing charts.
 */
window.addEventListener("resize", window.lava.redrawCharts);

lava.run();
</script>
<script type="text/javascript">
        lava.events.on('jsapi:ready', function (google) {
            /**
             * If the object does not exist for a given chart type, initialize it.
             * This will prevent overriding keys when multiple charts of the same
             * type are being rendered on the same page.
             */
            if ( typeof lava.charts.AreaChart == "undefined" ) {
                lava.charts.AreaChart = {};
            }

            //Creating a new lavachart object
            lava.charts.AreaChart["registrations"] = new lava.Chart();

            //Checking if output div exists
            if (! document.getElementById("registrations_div")) {
                throw new Error('[Lavacharts] No matching element was found with ID "registrations_div"');
            }

            lava.charts.AreaChart["registrations"].render = function (data) {
                var $this = lava.charts.AreaChart["registrations"];

                $this.data = new google.visualization.DataTable({"cols":[{"type":"string","label":"Date"},{"type":"number","label":"Activated"},{"type":"number","label":"Not activated"}],"rows":[]}, 0.6);

                $this.options = {"backgroundColor":{"fill":"#333333","stroke":"#868686","strokeWidth":3},"hAxis":{"textStyle":{"color":"#bbb"},"gridlines":{"color":"#bbb","count":7}},"vAxis":{"textStyle":{"color":"#bbb"},"gridlines":{"color":"#646464","count":5}},"lineWidth":4,"legend":{"position":"top","alignment":"center","textStyle":{"color":"#bbb"}},"chartArea":{"width":"90%"},"titleTextStyle":{"color":"#bbb"},"focusTarget":"category"};

                $this.chart = new google.visualization.AreaChart(document.getElementById("registrations_div"));

                

                

                $this.chart.draw($this.data, $this.options);
            };

            lava.charts.AreaChart["registrations"].setData = function (data) {
                var $this = lava.charts.AreaChart["registrations"];

                $this.data = new google.visualization.DataTable(data, 0.6);
            };

            lava.charts.AreaChart["registrations"].redraw = function () {
                var $this = lava.charts.AreaChart["registrations"];

                $this.chart.draw($this.data, $this.options);
            };

            lava.registerChart("AreaChart", "registrations");

            google.load('visualization', '1', {
                packages: ['corechart'],
                callback: function() {
                    lava.charts.AreaChart["registrations"].render();
                }
            });
        });
</script>            
                        
<script type="text/javascript">
        lava.events.on('jsapi:ready', function (google) {
            /**
             * If the object does not exist for a given chart type, initialize it.
             * This will prevent overriding keys when multiple charts of the same
             * type are being rendered on the same page.
             */
            if ( typeof lava.charts.AreaChart == "undefined" ) {
                lava.charts.AreaChart = {};
            }

            //Creating a new lavachart object
            lava.charts.AreaChart["loginAttempts"] = new lava.Chart();

            //Checking if output div exists
            if (! document.getElementById("loginAttempts_div")) {
                throw new Error('[Lavacharts] No matching element was found with ID "loginAttempts_div"');
            }

            lava.charts.AreaChart["loginAttempts"].render = function (data) {
                var $this = lava.charts.AreaChart["loginAttempts"];

                $this.data = new google.visualization.DataTable({"cols":[{"type":"string","label":"Date"},{"type":"number","label":"Successful Attempts"},{"type":"number","label":"Failed Attempts"},{"type":"number","label":"Failed because not activated"}],"rows":[{"c":[{"v":"January 31"},{"v":0},{"v":0},{"v":0}]},{"c":[{"v":"February 1"},{"v":0},{"v":0},{"v":0}]},{"c":[{"v":"February 2"},{"v":0},{"v":0},{"v":0}]},{"c":[{"v":"February 3"},{"v":0},{"v":0},{"v":0}]},{"c":[{"v":"February 4"},{"v":0},{"v":0},{"v":0}]},{"c":[{"v":"February 5"},{"v":0},{"v":0},{"v":0}]},{"c":[{"v":"February 6"},{"v":1},{"v":0},{"v":1}]}]}, 0.6);

                $this.options = {"backgroundColor":{"fill":"#333333","stroke":"#868686","strokeWidth":3},"hAxis":{"textStyle":{"color":"#bbb"},"gridlines":{"color":"#bbb","count":7}},"vAxis":{"textStyle":{"color":"#bbb"},"gridlines":{"color":"#646464","count":5}},"lineWidth":4,"legend":{"position":"top","alignment":"center","textStyle":{"color":"#bbb"}},"chartArea":{"width":"90%"},"titleTextStyle":{"color":"#bbb"},"focusTarget":"category"};

                $this.chart = new google.visualization.AreaChart(document.getElementById("loginAttempts_div"));

                

                

                $this.chart.draw($this.data, $this.options);
            };

            lava.charts.AreaChart["loginAttempts"].setData = function (data) {
                var $this = lava.charts.AreaChart["loginAttempts"];

                $this.data = new google.visualization.DataTable(data, 0.6);
            };

            lava.charts.AreaChart["loginAttempts"].redraw = function () {
                var $this = lava.charts.AreaChart["loginAttempts"];

                $this.chart.draw($this.data, $this.options);
            };

            lava.registerChart("AreaChart", "loginAttempts");

            google.load('visualization', '1', {
                packages: ['corechart'],
                callback: function() {
                    lava.charts.AreaChart["loginAttempts"].render();
                }
            });
        });
</script>            
```

**Comment by kevinkhill** [comment]
my first assumption is that google.load is trying to get corechart twice and causing hiccups, but I have made a test in the past of 100 linecharts on one page, and it rendered them all.

Do either load? neither load?

I noticed the first chart has no rows, is that intentional?

**Comment by kevinkhill** [comment]
What's also very strange is that I copied and pasted your javascript into a plain HTML file. Then I put the two divs for output (and added "https:" in the lava.js module because it was trying to load file://www.google.com/jsapi) and it rendered!

![issue-81_datatable-is-not-a-constructor](https://cloud.githubusercontent.com/assets/266076/12868459/83a1c388-ccbc-11e5-8a56-6ce32e6c573b.png)

**Comment by TheTechnoMan** [comment]
Hey, 

it is strange, sometimes they render perfectly fine and sometimes they don't. Very random :/

Am 6. Februar 2016 19:29:39 MEZ, schrieb Kevin Hill <notifications@github.com>:

**Comment by kevinkhill** [comment]
Well, that makes it tough to debug. Since the Javascript is making it to the page correctly and the charts rendered for me, I don't know how I can help figure this out. It has to either be something with the way my lava.js module handles everything, or something with Google

**Comment by TheTechnoMan** [comment]
Understandable. What would you need to debug it? Maybe you're up to a Teamviewer session? Or anything that I can do? Unfortunately I can't give out the code as this is an internal project.

**Comment by TheTechnoMan** [comment]
@kevinkhill I would really appreciate your help. Pretty much always get the error nowadays :(

**Comment by kevinkhill** [comment]
Seeing how this weekend is valentine's day, I probably won't be working, but maybe Monday? I could TeamView with you or we could get on gitter and chat. I'd like to sort this out, because I know how frustrating bugs can be

**Comment by TheTechnoMan** [comment]
That sounds great to me! Have a good valentine's day, Kevin! :)

**Comment by kevinkhill** [comment]
Just post here on Monday when you're available, I'll get an alert and reply if it's a good time.

**Comment by TheTechnoMan** [comment]
Alright, even though you're in a complete different timezone, I am always up that late anyways, so that should be fine. See you then!

**Comment by TheTechnoMan** [comment]
@kevinkhill I'm online for a bit, sadly not too long today cause I had no sleep. Thanks!

**Comment by kevinkhill** [comment]
What's a easier way we can chat, hangouts?

**Comment by TheTechnoMan** [comment]
I never used hangouts. How about a gitter chat and maybe teamviewer?

**Comment by TheTechnoMan** [comment]
Thanks for the session yesterday, I really hope that you can figure that out :+1:

**Comment by kevinkhill** [comment]
Sure thing, I want this to be bullet proof, so I am overhauling the lava.js module, trying to get to the bottom of this.

Good news though, I finally got the same error as you! No idea why! Hahahahaha *maniacal laughter*

**Comment by TheTechnoMan** [comment]
Your comment at the end made that post a totally different one LOL. Glad you get the error too, I mean, I guess. haha

Good luck and thanks again, really appreciated!

**Comment by kevinkhill** [comment]
So I think the problem is my semi-async flow of the lava.js module. I think that even though Google's jsapi is loading, some how the ajax calls happen sooner than expected? I don't know, trying to follow async workflows is challenging.

I am close (I think?) to a solution. I need to figure out how to
1. Output lava.js to the page (done)
2. Add the jsapi script tag to the page async (to not block page load) (done)
3. Once the jsapi is loaded, notify (I used events) each chart to render, since the google object should be available now.
4. Once each chart is done rendering, notify the lava.js module all charts are on page (brain hurting)
5. Once all events are received, continue and run my newly added lava.ready callback (to wrap the ajax calls in)

Promises? Events? Async... This is tricky, and I am not a pro javascripter

**Comment by kevinkhill** [comment]
If you are feeling adventurous, the 3.1 branch has (mostly) working code for ```$lava->renderAll()``` as well as a new ```lava.ready()``` method for (hopefully) protecting against race conditions.

I put lots of notes in the [changelog](https://github.com/kevinkhill/lavacharts/blob/master/CHANGELOG.md) about the improvements if you feel like taking it for a spin.

**Comment by TheTechnoMan** [comment]
Hey,

just changed to the master branch, now I get:

ErrorException in Volcano.php line 109: array_key_exists() expects exactly 2 parameters, 3 given

Is there something else that needs to be done on my side, after changing to master?

Thanks!

**Comment by kevinkhill** [comment]
Oh, no, that's my mistake. I'm realizing now that ```in_array()``` can be strict with the third param, not ```array_key_exists()```

**Comment by TheTechnoMan** [comment]
Oh I see. Can you maybe quickly fix this? Would love to test it before I go home (it's half past 3am here lol).

**Comment by kevinkhill** [comment]
Yup, patched

**Comment by TheTechnoMan** [comment]
lol, just did it myself, now I get:

ErrorException in Renderable.php line 43: No ElementId was set for AreaChart("registrations"), using "registrations".

**Comment by kevinkhill** [comment]
That's supposed to just be a notice, not error... But yes, check out the changelog in the repo.

**Comment by kevinkhill** [comment]
I meant, the element ID is now set in the chart creation

**Comment by TheTechnoMan** [comment]
Not quite sure that you mean with "element ID". Is that the id in which it should render the chart, e.g. the div?

**Comment by TheTechnoMan** [comment]
If you take this as an example:

@areachart('registrations', 'registrations_div')

The elemtid would be registrations, right? Would I then just remove it from the render menthod and just pass the div id there?

**Comment by kevinkhill** [comment]
I haven't worked on the blade directives yet.  I mostly just wanted you to test all my work to the lava.js module.

If you could, comment out the blade directives and use
```<?= \Lava::renderAll(); ?>```

**Comment by TheTechnoMan** [comment]
Weird, that results in some sort of a redirection loop..

**Comment by kevinkhill** [comment]
Yes, the element ID I keep referring to is the div's ID where the chart will be output into.

**Comment by TheTechnoMan** [comment]
Now I get:

Error: [Lavacharts] Chart with label "registrations" was not found.
TypeError: this._readyCallback is not a function

Not quite sure why

**Comment by kevinkhill** [comment]
Did you wrap your ajax in ```lava.ready()```?

**Comment by TheTechnoMan** [comment]
Oh, missed that. Now I get:

Error: [Lavacharts] object is not a valid callback.
TypeError: window.google.visualization is undefined
TypeError: this._readyCallback is not a function

**Comment by kevinkhill** [comment]
Haha, damn, I'm starting to think I asked for testing too soon. I just got excited it worked on my end.

Can you post the JS your ajax is doing? using:
```javascript
lava.ready(function() {
    $.getJSON(...
});
```
should be assigning this._readyCallback in the lava.js module...

**Comment by TheTechnoMan** [comment]
sure

```
            lava.ready(
                    $.getJSON('{{ route('get.graphs.fetch', 'registrations,loginAttempts') }}', function (dataTableJson) {
                        lava.loadData('registrations', dataTableJson.registrations, function (chart) {
                            $el_registrations.trigger('loading-overlay:hide');
                        });

                        lava.loadData('loginAttempts', dataTableJson.loginAttempts, function (chart) {
                            $el_loginAttempts.trigger('loading-overlay:hide');
                        });
                    })
            );
```

**Comment by kevinkhill** [comment]
You're missing the anonymous function in ```lava.ready```

Check out my last post

**Comment by TheTechnoMan** [comment]
Sorry my bad. I have some great new. IT WORKS! Thank you so much for your work! I leave you the honor to close this again, hopefully permantely this time!

Thanks!

**Comment by kevinkhill** [comment]
I've gone back and forth, because I am not confident with my git work flow, but right now, I'm treating master as my workspace. I  have been merging feature complete stuff into the numbered branches.

So, 3.1 will be stable (in the future) and master might work, might not.

**Comment by tahiryasin** [comment]
Hi @kevinkhill 
I hope you are doing great.

I am using your awesome package and wanted to get help with an issue. Actually `lava.ready` function never fires for me and I don't know why. Could you please see my code and let me know if I am missing something here?

here is my route file
```php

Route::get('/lava', function (){
	$temps = \Lava::DataTable();
	$temps->addStringColumn('Type')
		  ->addNumberColumn('Value');
	\Lava::GaugeChart('Battery', $temps, [
		'max'=>24,
		'width'      => 400,
		'greenFrom'  => 12,
		'greenTo'    => 24,
		'yellowFrom' => 5,
		'yellowTo'   => 12,
		'redFrom'    => 0,
		'redTo'      => 5,
		'majorTicks' => [
			'Critical',
			'Safe',
		]
	]);
	return view('lava');
});
```

here is my `lava.blade.php` layout file

```html
<!DOCTYPE html>
<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <!-- CSRF Token -->
    <meta name="csrf-token" content="{{ csrf_token() }}">

    <title>{{ config('app.name', 'Laravel') }}</title>

</head>
<body>
     @yield('content')
</body>
</html>
```

here is my `lava.blade.php` view file

```php
@extends('layouts.lava')

@section('content')

@gaugechart('Battery', 'batteryGuage')
<div id="batteryGuage"></div>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
<script>
lava.ready(function (){
	console.log('lava is ready');
	var dataURL = "{{route('batteryGuage')}}";
	$.getJSON( dataURL, {
		device_id:1
	})
	.done(function( dataTableJson ) {
		lava.loadData('Battery', dataTableJson);
	});
});
</script>
@endsection
```

**Comment by kevinkhill** [comment]
are you getting any javascript errors in the console?

**Comment by tahiryasin** [comment]
@kevinkhill Thanks for your kind response. I have no errors/warning in my console.

**Comment by tahiryasin** [comment]
@kevinkhill btw, I had posted this problem in a separate issue [https://github.com/kevinkhill/lavacharts/issues/272](https://github.com/kevinkhill/lavacharts/issues/272)

**Comment by BrandNewUsername** [comment]
I ran into the same error on my production server: TypeError: google.visualization.DataTable is not a constructor
The error never occurred on my development environment though. 

I have an API generating / showing the data. I cache the query results. The first time the graph is rendered (so without cached results and therefore pretty slow) the error never occurred. Only after the first time I render the graph, with cached query results, the error appears frequently.

Since the problem only occurs on my development environment, which had less performance than production, I decided to add a sleep(2) to my API before showing the results. After adding this, the error did not appear again.

Though it's not a solution, it's a work-around that might help.
